### ISSUE TITLE
Investigate and create ADR for underlying engine implementation

### ISSUE THREAD

**Issue by brandtkeller** [defenseunicorns-labs/lula1#39]
Focusing on Kubernetes

Need to develop edge-cases for using the Kyverno engine as a means to check configuration across the cluster.

There are instances were it may be difficult to validate a global configuration using the kyverno engine. 

IE we can validate that all PeerAuthentication resources for istio are set to a specific configuration - but we cannot easily validate that all applicable namespaces have a peerauthentication resource present.

This may look like:
- Switching to building an internal engine for querying
- Hybrid solution for doing some things (single resource) with kyverno and otherwise doing custom multi-resource validation
- Implementing a solution that leverages kyverno fully as the engine (maybe switch to a policy for all required namespaces instead of relying on clusterpolicy)

Develop an Architecture Decision record document for chosen implementation and intended path forward/implications.

**Comment by lucasrod16** [comment]
@brandtkeller you noted that "we cannot easily validate that all applicable namespaces have a peerauthentication resource present" with Kyverno. I intend to begin running through some experiments to test such edge cases this week, but I'm curious about a few things: 

- what kubernetes resources will we need to be able to query in a kubernetes cluster?
- do we expect to need to be able to query non-namespaced (global) resources such as CustomResourceDefinitions?
- should this engine/querying capability have global and namespaced permissions out of the box? or should it only have permissions that a user defines/allows for it to have with kubernetes RBAC resources (roles/ClusterRoles, serviceAccounts, etc.)?

**Comment by lucasrod16** [comment]
Making note that the questions asked in [this comment](https://github.com/defenseunicorns/lula/issues/39#issuecomment-1448629051) have been addressed in a synchronous meeting.

We will need to be able to query any/every resource in a kubernetes cluster. TBD on how to handle permissions

**Comment by lucasrod16** [comment]
Doing some testing/experimentation with Kyverno global cluster querying capabilities on [this branch](https://github.com/defenseunicorns/lula/tree/test-cluster-querying).

The `demo/` directory is where all the action is happening right now

**Comment by lucasrod16** [comment]
@brandtkeller what is the goal of being able to query namespaces for the existence of specific resources (peerauthentications, roles, serviceAccounts, etc)? If a resource is missing, but is expected to be there, would we fail validation? Or is there another scenario/outcome we're wanting here?

**Comment by brandtkeller** [comment]
> @brandtkeller what is the goal of being able to query namespaces for the existence of specific resources (peerauthentications, roles, serviceAccounts, etc)? If a resource is missing, but is expected to be there, would we fail validation? Or is there another scenario/outcome we're wanting here?

Validating the existence of a specific resource in a namespace **will** be a valid scenario when coupled with then performing configuration validation. If it doesn't exist, I expect the validation to fail. 

So to clarify - in the case of something like peerauthentication - that would likely be a two-part validation:
- Validate that there is at a minimum 1 in every namespace (exclusions are valid here)
- Validate it's configuration against how the control is satisfied.

**Comment by lucasrod16** [comment]
> Validating the existence of a specific resource in a namespace **will** be a valid scenario when coupled with then performing configuration validation. If it doesn't exist, I expect the validation to fail.
> 

When a resource isn't present in a given namespace, the current logic does produce a fail status and a compliance report with a result of `Fail`. To give some more context to this:

After creating a namespace and applying the `pod.fail.yaml` file to a cluster into that namespace, I then deleted the `demo-pod` from that namespace, ran `lula execute` against the oscal-component.yaml file in the demo/ directory.

Attached is a screenshot of the terminal output, and the generated compliance report, which is attached as a plain text file because github [doesn't support attaching yaml files](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/attaching-files). The oscal-component.yaml file from the `main` branch was used  with `lula execute` with no modifications.

<img width="674" alt="Screenshot 2023-02-28 at 8 56 56 PM" src="https://user-images.githubusercontent.com/87675701/222033723-90ab7601-1b18-459b-a932-cc6cb7f04867.png">

[compliance_report.txt](https://github.com/defenseunicorns/lula/files/10856773/compliance_report.txt)

Given these artifacts, where do we see gaps in the generated outputs? The reason I ask is because, in a sense, validation is technically failing when a resource doesn't exist in a given namespace. I think it would be helpful to continue to flesh out what failing validation means for end-users. Should the generated compliance report supply different/more informative data? Or, since the resource didn't exist, should a compliance report even be generated?

If a resource doesn't exist, is that the compliance violation itself in this case? Or are we saying that we cannot perform proper compliance validation if the resource doesn't exist and logging a warning or error?

**Comment by brandtkeller** [comment]
You're into the thick of it now and on the right path!

I agree with your sentiment around continue to define what information is presented to the user (keeping in mind the bounds of what will be produced in the reports that are placed before AO's etc).

I believe there are two layers to this discussion - please correct me if I misunderstood

1. If (as described above) the resource for which validation was performed was missing - we must define if this is actually a failure (maybe it is still considering the declaration?) I'd argue that if we wrote a CIS benchmark OSCAL document - that there should be policies that target 0 resources as likely some are configuring some known mis-configured resources that are not present. To summarrize here - there is likely more data to digest here to support the future of a pass/fail with consideration to the control (maybe this is some additional metadata?). Note: [the pass/fail logic](https://github.com/defenseunicorns/lula/blob/main/src/cmd/execute/execute.go#L200) is pretty simple in nature and will need iteration.

2. More to the problem that I was originally testing. We should (and can) write a cluster policy that uses the external datasource to query and ensure the number of peerauthentications > 0 for each namespace. That would be one validation that fails when checking for a simple count is 0 (does not exist).

let me know if that shines more light and answers your question or not.

**Comment by lucasrod16** [comment]
That helps a lot, thanks!

**Comment by lucasrod16** [comment]
@brandtkeller Since the branch I'm working on will probably be short-lived, I'll share notes here.

[notes.md](https://github.com/defenseunicorns/lula/files/10866015/notes.md)

I was able to verify that Kyverno is able to validate that all applicable/specified namespaces have a peerauthentication resource present.

I modified the kyverno rules spec in `./demo/oscal-component.yaml` to be:

```yaml
rules:
- name: istio-controlplane_AC-4(peer-auth-for-every-namespace)
  match:
    any:
    - resources:
        kinds:
        - "Namespace"
        names:
        - "foo"
        - "test"
        - "test1"
        - "test2"
  context:
  - name: peerAuthCount
    apiCall:
      urlPath: "apis/security.istio.io/v1beta1/peerauthentications"
      jmesPath: "items | length(@)"
  validate:
    message: "The specified namespaces must have at least 1 PeerAuthentication resource."
    deny:
      conditions:
        any:
        - key: "{{peerAuthCount}}"
          operator: LessThan
          value: 1
```

Very similar to the configuration mentioned in [this issue](https://github.com/defenseunicorns/lula/issues/19).

Some notable changes:

- the API endpoint path changed from `apis/security.istio.io/v1beta1/namespaces/{{request.namespace}}/peerauthentications` to `apis/security.istio.io/v1beta1/peerauthentications`

  - removing the `namespaces` from the path means the API call sent to the kubernetes API server queries all namespaces rather than filtering down to a specific namespace
  
  - the Kind has been changed from `Pod` to `Namespace` to target namespace resources
  
  - it is targeting namespaces by name with the `names:` field
  
  Unless I am misunderstanding the original question, I think this answers whether we can "validate that all applicable namespaces have a peerauthentication resource present" with Kyverno or not. We can.
  
 Are there other edge cases we want to/should test for in regards to Kyverno's functionality for our use cases?

**Comment by lucasrod16** [comment]
The notes.md file I attached in the previous comment walks through the steps I took to reproduce the results

**Comment by lucasrod16** [comment]
It actually may be incorrect that removing namespaces from the API endpoint path means it queries all namespaces since in this case we are querying for namespaces themselves. Need to dig more into that. It's somehow querying the global state of a cluster

**Comment by brandtkeller** [comment]
This looks like a valid test to start to define and understand when something passes or fails. Part of our previous problem was documented in #19 and may provide some insights worth testing. Particularly, we know kyverno can match on a resource in 1->n namespaces. Can we write policies that allow comparing two resources? do we even need to do that if we can just split out the validation into multiple policies and keep each focused on one resource? just some high-level food for thought. 

On the above, I'd be testing pass/fail and ensuring we can target specific namespaces (or that all that are not excluded were targeted). I'd also be thinking about static-analysis - Would this validation run against Kubernetes yaml? Another datapoint for the decision - as maybe we remove validations that require querying the API server during execution of static manifests? (as you can see, I don't yet know the answer to much of this either).

I'll keep stewing on other potential edge-cases.

**Comment by lucasrod16** [comment]
I must have misunderstood the original intent if we were already aware that kyverno can check if particular namespaces have peerauth resources or not.

I agree that testing static-analysis will be important since we will likely want to be able to do pre-flight validation of manifests, as well as runtime validation in a cluster.

I am not entirely clear on what is meant by "Can we write policies that allow comparing two resources?". Is this referring to whether we can match against multiple kubernetes resources (pods, secrets, etc) in a single policy to validate with kyverno? It's entirely likely there's a concept of comparing resources that I'm not aware of with regards to policy/compliance

**Comment by brandtkeller** [comment]
In #19 we were trying to write a validation that could be enforced as well. IE - how do you enforce blocking pod creation before the proper peerAuthentications have been applied - If I also have a clusterPolicy for peerAuthentications and ensuring they are mtls strict - then both together could reasonably enforce this idea.


The other target I had for assisting in the determination is either (or both):
- break-down the istio OSCAL and start to write suedo statements about validation - translate these to the various engines
- Look at the other OSCAL component definitions available and start to write the validations that can be performed

**Comment by brandtkeller** [comment]
Closed as of #71 - selection of multi-provider support.
